File: plugin/bukkit/src/main/java/cn/fd/fdutilities/util/ClassUtil.java
package cn.fd.fdutilities.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.util.List;
import java.util.jar.JarEntry;
import java.util.jar.JarInputStream;

public class ClassUtil {

    @Nullable
    public static <T> Class<? extends T> findClass(@NotNull final File file, @NotNull final Class<T> clazz) throws IOException, ClassNotFoundException {
        //检查文件是否存在，如果不存在，就返回空
        if (!file.exists()) {
            return null;
        }

        final URL jar = file.toURI().toURL();
        final URLClassLoader loader = new URLClassLoader(new URL[]{jar}, clazz.getClassLoader());
        final List<String> matches = new ArrayList<>();
        final List<Class<? extends T>> classes = new ArrayList<>();

        try (final JarInputStream stream = new JarInputStream(jar.openStream())) {
            JarEntry entry;
            while ((entry = stream.getNextJarEntry()) != null) {
                final String name = entry.getName();
                if (!name.endsWith(".class")) {
                    continue;
                }

                matches.add(name.substring(0, name.lastIndexOf('.')).replace('/', '.'));
            }

            for (final String match : matches) {
                try {
                    final Class<?> loaded = loader.loadClass(match);
                    if (clazz.isAssignableFrom(loaded)) {
                        classes.add(loaded.asSubclass(clazz));
                    }
                } catch (final NoClassDefFoundError ignored) {
                }
            }
        }
        if (classes.isEmpty()) {
            loader.close();
            return null;
        }
        return classes.get(0);
    }

}

File: project/core/src/main/kotlin/cn/fd/utilities/util/ClassUtil.kt
package cn.fd.utilities.util

import java.io.File
import java.io.IOException
import java.net.URLClassLoader
import java.util.*
import java.util.jar.JarEntry
import java.util.jar.JarFile

object ClassUtil {

    @Throws(IOException::class, ClassNotFoundException::class)
    fun <T> findClass(file: File, clazz: Class<T>): Class<out T>? {
        //检查文件是否存在
        if (!file.exists()) {
            return null
        }

        val jar = file.toURI().toURL()
        val loader = URLClassLoader(arrayOf(jar), clazz.classLoader)
        val classes: MutableList<Class<out T>> = ArrayList()

        val entries: MutableList<JarEntry> = arrayListOf()

        JarFile(file).entries().let { e ->
            while (e.hasMoreElements()) {
                e.nextElement().let {
                    if (it.name.endsWith(".class")) entries.add(it)
                }
            }
        }

        entries.forEach {
            val className = it.name.replace('/', '.').substring(0, it.name.length - 6)
            try {
                //val loaded = loadClass(name)
                val loaded = loader.loadClass(className)
                if (clazz.isAssignableFrom(loaded)) {
                    classes.add(loaded.asSubclass(clazz))
                }
            } catch (ignored: NoClassDefFoundError) {
            }
        }

        if (classes.isEmpty()) {
            return null
        }
        return classes[0]
    }


    /**
     * 加载类
     */
    @Throws(ClassNotFoundException::class)
    fun loadClass(className: String, isInitialized: Boolean = false): Class<*> {
        return Class.forName(className, isInitialized, Thread.currentThread().contextClassLoader)
    }


}
